In [1]:
from shapely.geometry import Point
import contextily as ctx
from shapely.geometry import Point
from typing import List, Union, Set
import os
import requests
import geopandas as gpd
import pandas as pd
import warnings
from sqlalchemy import create_engine
from sqlalchemy.sql import text
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.animation as animation
from pandas import DataFrame
from matplotlib.animation import FuncAnimation
from IPython.display import HTML
warnings.filterwarnings("ignore")

In [2]:

def read_shapefile(shapefile_path: str) -> gpd.GeoDataFrame:
    """
    Reads the shapefile into a GeoDataFrame.

    Args:
    - shapefile_path (str): Path to the shapefile.

    Returns:
    - gpd.GeoDataFrame: The GeoDataFrame read from the shapefile.
    """
    return gpd.read_file(shapefile_path)

def filter_columns(gdf: gpd.GeoDataFrame, columns: List[str]) -> gpd.GeoDataFrame:
    """
    Filters the GeoDataFrame to include only specified columns.

    Args:
    - gdf (gpd.GeoDataFrame): The original GeoDataFrame.
    - columns (List[str]): A list of column names to retain.

    Returns:
    - gpd.GeoDataFrame: The GeoDataFrame with only the specified columns.
    """
    return gdf[columns]

def remove_duplicates(gdf: gpd.GeoDataFrame, subset: str) -> gpd.GeoDataFrame:
    """
    Removes duplicate rows based on a specified subset of columns.

    Args:
    - gdf (gpd.GeoDataFrame): The GeoDataFrame to process.
    - subset (str): Column name to check for duplicates.

    Returns:
    - gpd.GeoDataFrame: The GeoDataFrame with duplicates removed.
    """
    return gdf.drop_duplicates(subset=[subset])

def filter_invalid_zipcodes(gdf: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
    """
    Keeps only rows with valid 5-digit zipcodes.

    Args:
    - gdf (gpd.GeoDataFrame): The GeoDataFrame to process.

    Returns:
    - gpd.GeoDataFrame: The GeoDataFrame with only valid 5-digit zipcodes.
    """
    gdf['zipcode'] = gdf['zipcode'].astype(str)
    return gdf[gdf['zipcode'].str.isdigit() & (gdf['zipcode'].str.len() == 5)]

def process_zipcode_shapefile(shapefile_path: str) -> gpd.GeoDataFrame:
    """
    Cleans and prepares a zipcode shapefile for analysis.

    Args:
    - shapefile_path (str): Path to the zipcode shapefile.

    Returns:
    - gpd.GeoDataFrame: GeoDataFrame with processed zipcode data.
    """
    zipcode_gdf = read_shapefile(shapefile_path)
    essential_columns = ['ZIPCODE', 'geometry']
    zipcode_gdf = filter_columns(zipcode_gdf, essential_columns)
    zipcode_gdf = remove_duplicates(zipcode_gdf, 'ZIPCODE')
    zipcode_gdf.dropna(subset=essential_columns, inplace=True)
    zipcode_gdf.rename(columns={'ZIPCODE': 'zipcode'}, inplace=True)
    zipcode_gdf = filter_invalid_zipcodes(zipcode_gdf)
    common_crs = "EPSG:3857"
    zipcode_gdf.to_crs(common_crs, inplace=True)
    zipcode_gdf.columns = map(str.lower, zipcode_gdf.columns)

    return zipcode_gdf

def lat_validation(latitude):
    if not isinstance(latitude, (int, float)):
        raise TypeError("The latitude should be a float or int type")
    return -90 <= latitude <= 90


def long_validation(longitude: float) -> bool:
    if not isinstance(longitude, (int, float)):
        raise TypeError("The longitude should be a float or int type")
    return -180 <= longitude <= 180



In [3]:
geodf_zip_data = process_zipcode_shapefile("data/nyc_zipcodes/nyc_zipcodes.shp")
nyc_zips = geodf_zip_data['zipcode'].tolist()
nyc_zips = [float(element) for element in nyc_zips]

In [4]:

columns_needed = ['Unique Key', 'Created Date', 'Complaint Type', 'Incident Zip', 'Latitude', 'Longitude', 'Location']
def filter_t311(df: pd.DataFrame, column_needed: List[str], nyc_zip: Union[Set[str], List[str]]) -> gpd.GeoDataFrame:
    
    df.columns = ['Unique Key', 'Created Date', 'Closed Date', 'Agency', 'Agency Name',
       'Complaint Type', 'Descriptor', 'Location Type', 'Incident Zip',
       'Incident Address', 'Street Name', 'Cross Street 1', 'Cross Street 2',
       'Intersection Street 1', 'Intersection Street 2', 'Address Type',
       'City', 'Landmark', 'Facility Type', 'Status', 'Due Date',
       'Resolution Description', 'Resolution Action Updated Date',
       'Community Board', 'BBL', 'Borough', 'X Coordinate (State Plane)',
       'Y Coordinate (State Plane)', 'Open Data Channel Type',
       'Park Facility Name', 'Park Borough', 'Vehicle Type',
       'Taxi Company Borough', 'Taxi Pick Up Location', 'Bridge Highway Name',
       'Bridge Highway Direction', 'Road Ramp', 'Bridge Highway Segment',
       'Latitude', 'Longitude', 'Location']
    # Filter the DataFrame to only include necessary columns and drop rows with NaN values
    filtered = df[column_needed].dropna()

    # Further filter the DataFrame to only include rows where 'Incident Zip' is in nyc_zip
    filtered = filtered[filtered['Incident Zip'].isin(nyc_zip)]

    # Converting 'Created Date' to datetime
    filtered['Created Date'] = pd.to_datetime(filtered['Created Date'])

    # Define your date range
    start_date = pd.to_datetime('2015-01-01')
    end_date = pd.to_datetime('2023-09-30')

    # Filter the DataFrame for dates within the range
    filtered = filtered[(filtered['Created Date'] >= start_date) & (filtered['Created Date'] <= end_date)]

    # Apply latitude and longitude validation
    filtered = filtered[filtered['Latitude'].apply(lat_validation) & filtered['Longitude'].apply(long_validation)]

    # Convert to GeoDataFrame
    filtered = gpd.GeoDataFrame(filtered, geometry=gpd.points_from_xy(filtered['Longitude'], filtered['Latitude']))
    filtered.set_crs("EPSG:4326", inplace=True)
    filtered.to_crs("EPSG:3857", inplace=True)

    return filtered

In [5]:
## cehck
def filter_stc(df: pd.DataFrame, column_needed: List[str], nyc_zip: Set[str]) -> gpd.GeoDataFrame:
    
    df.columns =['created_at', 'tree_id', 'block_id', 'the_geom', 'tree_dbh',
       'stump_diam', 'curb_loc', 'status', 'health', 'spc_latin', 'spc_common',
       'steward', 'guards', 'sidewalk', 'user_type', 'problems', 'root_stone',
       'root_grate', 'root_other', 'trnk_wire', 'trnk_light', 'trnk_other',
       'brnch_ligh', 'brnch_shoe', 'brnch_othe', 'address', 'zipcode',
       'zip_city', 'cb_num', 'borocode', 'boroname', 'cncldist', 'st_assem',
       'st_senate', 'nta', 'nta_name', 'boro_ct', 'state', 'Latitude',
       'longitude', 'x_sp', 'y_sp']
    # Filter the DataFrame to only include necessary columns and drop rows with NaN values
    filtered = df[columns_needed].dropna()

    # Further filter the DataFrame to only include rows where 'zipcode' is in nyc_zip
    filtered = filtered[filtered['zipcode'].isin(nyc_zip)]

    # Converting 'created_at' to datetime
    filtered['created_at'] = pd.to_datetime(filtered['created_at'])

    # Define your date range
    start_date = pd.to_datetime('01/01/2015')
    end_date = pd.to_datetime('09/30/2023')  # Corrected date

    # Filter the DataFrame for dates within the range
    filtered = filtered[(filtered['created_at'] >= start_date) & (filtered['created_at'] <= end_date)]

    # Apply latitude and longitude validation
    filtered = filtered[filtered['Latitude'].apply(lat_validation) & filtered['longitude'].apply(long_validation)]

    # Convert to GeoDataFrame
    filtered = gpd.GeoDataFrame(filtered, geometry=gpd.points_from_xy(filtered['longitude'], filtered['Latitude']))
    filtered.set_crs("EPSG:4326", inplace=True)
    filtered.to_crs("EPSG:3857", inplace=True)

    return filtered

In [6]:

def filter_zillow(df: pd.DataFrame, nyc_zip: list) -> pd.DataFrame:
    # Selecting the required columns. Assuming the first column is 'RegionName' and the 9th to last are dates
    useful_cols = df.columns[9:].to_list() + ['RegionName']+['RegionID']
    filtered = df[useful_cols]

    # Drop rows where 'RegionName' is NaN
    filtered = filtered.dropna(subset=['RegionName'])
    filtered = filtered.drop_duplicates()

    # Filter rows where 'RegionName' is in the list of NYC zip codes
    filtered = filtered[filtered['RegionName'].isin(nyc_zip)]
    

    # Melting the DataFrame
    melted_df = filtered.melt(id_vars=['RegionID','RegionName'], value_vars=df.columns[9:], var_name='date', value_name='rent')
    melted_df = melted_df.reset_index()
    return melted_df


In [7]:
a = pd.read_csv("data/zillow_rent_data.csv")
a

RegionID  SizeRank  RegionName RegionType StateName State  \
0        91982         1       77494        zip        TX    TX   
1        91940         3       77449        zip        TX    TX   
2        91733         5       77084        zip        TX    TX   
3        93144         6       79936        zip        TX    TX   
4        62093         7       11385        zip        NY    NY   
...        ...       ...         ...        ...       ...   ...   
6717    418163     30158       89158        zip        NV    NV   
6718     72017     30490       32461        zip        FL    FL   
6719     58956     30490        2876        zip        RI    RI   
6720     91179     30490       76005        zip        TX    TX   
6721     61618     30490       10004        zip        NY    NY   

                   City                                   Metro  \
0                  Katy    Houston-The Woodlands-Sugar Land, TX   
1                  Katy    Houston-The Woodlands-Sugar Land, TX   
2               Houston    Houston-The Woodlands-Sugar Land, TX   
3               El Paso                             El Paso, TX   
4              New York   New York-Newark-Jersey City, NY-NJ-PA   
...                 ...                                     ...   
6717          Las Vegas        Las Vegas-Henderson-Paradise, NV   
6718  Panama City Beach  Crestview-Fort Walton Beach-Destin, FL   
6719   North Smithfield               Providence-Warwick, RI-MA   
6720          Arlington         Dallas-Fort Worth-Arlington, TX   
6721           New York   New York-Newark-Jersey City, NY-NJ-PA   

             CountyName   2015-01-31  ...   2022-12-31   2023-01-31  \
0      Fort Bend County  1606.206406  ...  1994.653463  2027.438438   
1         Harris County  1257.814660  ...  1749.697900  1738.217986   
2         Harris County          NaN  ...  1701.217520  1706.900064   
3        El Paso County          NaN  ...  1419.480272  1458.063897   
4         Queens County          NaN  ...  2935.808220  2895.699421   
...                 ...          ...  ...          ...          ...   
6717       Clark County          NaN  ...  3281.330738  3509.210744   
6718      Walton County          NaN  ...          NaN          NaN   
6719  Providence County          NaN  ...          NaN          NaN   
6720     Tarrant County          NaN  ...  2148.224601  2169.143026   
6721    New York County          NaN  ...  4006.972903  4000.923287   

       2023-02-28   2023-03-31   2023-04-30   2023-05-31   2023-06-30  \
0     2042.237444  2049.325559  2016.531345  2023.438976  2031.558202   
1     1747.305840  1758.407295  1758.891075  1762.980879  1771.751591   
2     1706.067787  1723.722320  1735.484670  1752.132904  1756.990323   
3     1471.726681  1466.734658  1456.175660  1462.478506  1466.267391   
4     2873.209025  2881.906361  2913.546218  2963.964134  3005.735342   
...           ...          ...          ...          ...          ...   
6717  3407.499896  3438.041504  3436.371804  3524.703410  3426.708975   
6718          NaN          NaN          NaN          NaN  2583.675563   
6719          NaN          NaN          NaN          NaN          NaN   
6720  2179.393248  2226.624684  2369.532530  2374.713926  2414.638428   
6721  4002.584212  4085.513015  4224.569373  4240.040733  4286.776061   

       2023-07-31   2023-08-31   2023-09-30  
0     2046.144009  2053.486247  2055.771355  
1     1779.338402  1795.384582  1799.631140  
2     1754.429516  1757.602011  1755.031490  
3     1490.237063  1488.180414  1494.366097  
4     3034.413822  3064.476503  3079.585783  
...           ...          ...          ...  
6717  3412.249969  3310.302151  3448.166667  
6718  2590.977335  2639.938102  2702.500000  
6719          NaN          NaN  2250.000000  
6720  2389.749852  2383.185013  2313.944444  
6721  4270.158740  4353.055657  4355.328283  

[6722 rows x 114 columns]

In [8]:
filter_zillow(a,nyc_zips)

index  RegionID  RegionName        date         rent
0          0     62093       11385  2015-01-31          NaN
1          1     62019       11208  2015-01-31          NaN
2          2     62046       11236  2015-01-31          NaN
3          3     61807       10467  2015-01-31          NaN
4          4     62085       11373  2015-01-31          NaN
...      ...       ...         ...         ...          ...
15325  15325     61773       10282  2023-09-30  7347.458333
15326  15326     62010       11109  2023-09-30  4529.358974
15327  15327     61620       10006  2023-09-30  4060.096154
15328  15328     61723       10162  2023-09-30  5011.666667
15329  15329     61618       10004  2023-09-30  4355.328283

[15330 rows x 5 columns]

In [60]:
def filter_bedbug(df: pd.DataFrame, column_needed: List[str], nyc_zip: Set[str]) -> gpd.GeoDataFrame:
    df.columns = ['Building ID', 'Registration ID', 'Borough', 'House Number',
       'Street Name', 'Postcode', '# of Dwelling Units',
       'Infested Dwelling Unit Count', 'Eradicated Unit Count',
       'Re-infested  Dwelling Unit Count', 'Filing Date',
       'Filing Period Start Date', 'Filling Period End Date', 'Latitude',
       'Longitude', 'Community Board', 'Council District', '2010 Census Tract',
       'BIN', 'BBL', 'NTA']
    # Ensure 'Postcode' and 'Filing Date' are in the needed columns
    if 'Postcode' not in column_needed or 'Filing Date' not in column_needed:
        raise ValueError("Required columns 'Postcode' and 'Filing Date' are missing.")

    # Selecting the required columns and drop rows with NaN values
    filtered = df[column_needed].dropna()
    filtered['Postcode']=filtered['Postcode'].astype("int")
    # Further filter the DataFrame to only include rows where 'Postcode' is in nyc_zip
    filtered = filtered[filtered['Postcode'].isin(nyc_zip)]

    # Converting 'Filing Date' to datetime
    filtered['Filing Date'] = pd.to_datetime(filtered['Filing Date'])

    # Define your date range
    start_date = pd.to_datetime('01/01/2015')
    end_date = pd.to_datetime('09/30/2023')

    # Filter the DataFrame for dates within the range
    filtered = filtered[(filtered['Filing Date'] >= start_date) & (filtered['Filing Date'] <= end_date)]
    filtered = filtered.reset_index()
    # Convert to GeoDataFrame (assuming Latitude and Longitude columns are present)
    if 'Latitude' in filtered.columns and 'Longitude' in filtered.columns:
        gdf = gpd.GeoDataFrame(filtered, geometry=gpd.points_from_xy(filtered['Longitude'], filtered['Latitude']))

        # Set a CRS for the GeoDataFrame
        # Replace 'EPSG:3857' with the correct EPSG code to match your PostGIS table
        gdf.set_crs(epsg=3857, inplace=True)
        return gdf
    else:
        raise ValueError("Latitude and Longitude columns are required for GeoDataFrame conversion.")

# Example usage:

In [61]:
chunk_size = 100000

# Initialize lists to hold processed chunks

stc_chunks = []
zillow_chunks = []
bedbug_chunks = []

In [11]:
# Process and store chunks for 'StreetTreesCensus_TREES'
for chunk in pd.read_csv('data/2015StreetTreesCensus_TREES.csv', chunksize=chunk_size):
    columns_needed = ['created_at', 'Latitude', 'longitude', 'tree_id', 'zipcode', 'health', 'spc_common']
    processed_chunk = filter_stc(chunk, columns_needed, nyc_zip=nyc_zips)
    stc_chunks.append(processed_chunk)
geodf_tree_data = pd.concat(stc_chunks)



In [12]:
# Load shapefile and save to the database
import geopandas as gpd

# Load your shapefile
geodf_zipcode_data = gpd.read_file('data/nyc_zipcodes/nyc_zipcodes.shp')

# Selecting only the 'ZIPCODE' and 'geometry' columns
geodf_zipcode_data = geodf_zipcode_data[['ZIPCODE', 'geometry']]

# Set the initial CRS to EPSG:2263 if it's not already set
geodf_zipcode_data.set_crs(epsg=2263, inplace=True)

# Transform the CRS from EPSG:2263 to EPSG:3857
geodf_zipcode_data = geodf_zipcode_data.to_crs(epsg=3857)
geodf_zipcode_data = geodf_zipcode_data.drop_duplicates(subset=['ZIPCODE'])


In [13]:
# Process and store chunks for 'zillow_rent_data'
for chunk in pd.read_csv('data/zillow_rent_data.csv', chunksize=chunk_size):
    processed_chunk = filter_zillow(chunk, nyc_zip=nyc_zips)
    zillow_chunks.append(processed_chunk)
df_zillow_data = pd.concat(zillow_chunks)



In [62]:
# Process and store chunks for 'Bedbug_Reporting'
for chunk in pd.read_csv('data/Bedbug_Reporting_20231203.csv', chunksize=chunk_size):
    columns_needed = ['Building ID', 'Postcode', 'Filing Date', 'Eradicated Unit Count', 'Re-infested  Dwelling Unit Count','Latitude','Longitude']
    processed_chunk = filter_bedbug(chunk, columns_needed, nyc_zip=nyc_zips)
    bedbug_chunks.append(processed_chunk)
df_bedbug_data = pd.concat(bedbug_chunks)

In [21]:
# Specify the chunk size
t311_chunks = []

# Process and store chunks for '311_Service_Requests'
for chunk in pd.read_csv('data/311_Service_Requests_from_2010_to_Present_20231129.csv', chunksize=chunk_size):
    columns_needed = ['Unique Key', 'Created Date', 'Complaint Type', 'Incident Zip', 'Latitude', 'Longitude', 'Location']
    processed_chunk = filter_t311(chunk, columns_needed, nyc_zip=nyc_zips)
    t311_chunks.append(processed_chunk)
geodf_311_data = pd.concat(t311_chunks)


In [22]:
# Show basic info about each dataframe
geodf_zipcode_data.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
Index: 248 entries, 0 to 262
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   ZIPCODE   248 non-null    object  
 1   geometry  248 non-null    geometry
dtypes: geometry(1), object(1)
memory usage: 5.8+ KB


In [23]:
# Show first 5 entries about each dataframe
geodf_zipcode_data.head()

ZIPCODE                                           geometry
0   11436  POLYGON ((-8216029.470 4965682.769, -8216011.9...
1   11213  POLYGON ((-8230673.455 4965216.008, -8230392.3...
2   11212  POLYGON ((-8226837.796 4963911.170, -8226758.2...
3   11225  POLYGON ((-8232963.912 4963884.338, -8232717.3...
4   11218  POLYGON ((-8234534.400 4960940.544, -8234516.0...

In [24]:
geodf_311_data.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
Index: 19436579 entries, 2796244 to 33897229
Data columns (total 8 columns):
 #   Column          Dtype         
---  ------          -----         
 0   Unique Key      int64         
 1   Created Date    datetime64[ns]
 2   Complaint Type  object        
 3   Incident Zip    object        
 4   Latitude        float64       
 5   Longitude       float64       
 6   Location        object        
 7   geometry        geometry      
dtypes: datetime64[ns](1), float64(2), geometry(1), int64(1), object(3)
memory usage: 1.3+ GB


In [25]:
geodf_311_data.head()

Unique Key        Created Date                   Complaint Type  \
2796244    22333067 2020-04-15 14:17:16  Construction Safety Enforcement   
6967776    27462375 2015-10-01 00:00:00   Unsanitary Animal Pvt Property   
7064133    27684062 2015-10-01 00:00:00   Unsanitary Animal Pvt Property   
7184490    27814302 2015-10-01 00:00:00   Unsanitary Animal Pvt Property   
7314518    29672793 2015-01-06 07:17:31                 Street Condition   

        Incident Zip   Latitude  Longitude  \
2796244      11205.0  40.697312 -73.964802   
6967776      11414.0  40.660428 -73.841226   
7064133      10301.0  40.633330 -74.105250   
7184490      11417.0  40.684300 -73.837892   
7314518      11235.0  40.593575 -73.932814   

                                         Location  \
2796244   (40.69731208999635, -73.96480244651633)   
6967776   (40.66042792940118, -73.84122633877455)   
7064133   (40.63332969998893, -74.10525005425679)   
7184490   (40.68430012358396, -73.83789200050686)   
7314518  (40.593575191824996, -73.93281445608707)   

                                 geometry  
2796244  POINT (-8233724.145 4967797.263)  
6967776  POINT (-8219967.716 4962383.151)  
7064133  POINT (-8249358.701 4958407.393)  
7184490  POINT (-8219596.539 4965886.935)  
7314518  POINT (-8230163.258 4952577.668)

In [26]:
geodf_tree_data.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
Index: 652167 entries, 0 to 683787
Data columns (total 8 columns):
 #   Column      Non-Null Count   Dtype         
---  ------      --------------   -----         
 0   created_at  652167 non-null  datetime64[ns]
 1   Latitude    652167 non-null  float64       
 2   longitude   652167 non-null  float64       
 3   tree_id     652167 non-null  int64         
 4   zipcode     652167 non-null  int64         
 5   health      652167 non-null  object        
 6   spc_common  652167 non-null  object        
 7   geometry    652167 non-null  geometry      
dtypes: datetime64[ns](1), float64(2), geometry(1), int64(2), object(2)
memory usage: 44.8+ MB


In [27]:
geodf_tree_data.head()

created_at   Latitude  longitude  tree_id  zipcode health       spc_common  \
0 2015-08-27  40.723092 -73.844215   180683    11375   Fair        red maple   
1 2015-09-03  40.794111 -73.818679   200540    11357   Fair          pin oak   
2 2015-09-05  40.717581 -73.936608   204026    11211   Good      honeylocust   
3 2015-09-05  40.713537 -73.934456   204337    11211   Good      honeylocust   
4 2015-08-30  40.666778 -73.975979   189565    11215   Good  American linden   

                           geometry  
0  POINT (-8220300.436 4971583.163)  
1  POINT (-8217457.809 4982020.303)  
2  POINT (-8230585.520 4970773.712)  
3  POINT (-8230346.012 4970179.889)  
4  POINT (-8234968.356 4963315.009)

In [28]:
df_zillow_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15330 entries, 0 to 15329
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   index       15330 non-null  int64  
 1   RegionID    15330 non-null  int64  
 2   RegionName  15330 non-null  int64  
 3   date        15330 non-null  object 
 4   rent        9041 non-null   float64
dtypes: float64(1), int64(3), object(1)
memory usage: 599.0+ KB


In [29]:
df_zillow_data.head()

index  RegionID  RegionName        date  rent
0      0     62093       11385  2015-01-31   NaN
1      1     62019       11208  2015-01-31   NaN
2      2     62046       11236  2015-01-31   NaN
3      3     61807       10467  2015-01-31   NaN
4      4     62085       11373  2015-01-31   NaN

In [63]:
df_bedbug_data.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
Index: 364683 entries, 0 to 68765
Data columns (total 9 columns):
 #   Column                            Non-Null Count   Dtype         
---  ------                            --------------   -----         
 0   index                             364683 non-null  int64         
 1   Building ID                       364683 non-null  int64         
 2   Postcode                          364683 non-null  int64         
 3   Filing Date                       364683 non-null  datetime64[ns]
 4   Eradicated Unit Count             364683 non-null  float64       
 5   Re-infested  Dwelling Unit Count  364683 non-null  float64       
 6   Latitude                          364683 non-null  float64       
 7   Longitude                         364683 non-null  float64       
 8   geometry                          364683 non-null  geometry      
dtypes: datetime64[ns](1), float64(4), geometry(1), int64(3)
memory usage: 27.8 MB


In [31]:
df_bedbug_data.head()

index  Building ID  Postcode Filing Date  Eradicated Unit Count  \
0      0       344156   11222.0  2021-02-16                    0.0   
1      1       170777   11204.0  2022-03-04                    0.0   
2      2       358122   11216.0  2023-07-10                    0.0   
3      3       705843   11385.0  2021-05-05                    0.0   
4      4       877716   11212.0  2023-07-06                    0.0   

   Re-infested  Dwelling Unit Count   Latitude  Longitude  \
0                               0.0  40.727680 -73.949030   
1                               0.0  40.613333 -73.994291   
2                               0.0  40.684545 -73.945667   
3                               0.0  40.702903 -73.902286   
4                               0.0  40.656147 -73.903403   

                 geometry  
0  POINT (-73.949 40.728)  
1  POINT (-73.994 40.613)  
2  POINT (-73.946 40.685)  
3  POINT (-73.902 40.703)  
4  POINT (-73.903 40.656)

## Part 2: Storing Data

In [35]:
# this code is not be able to use 
!createdb e4501project
!psql --dbname e4501project -c 'CREATE EXTENSION if NOT EXISTS postgis;'

CREATE EXTENSION


### Creating Tables


These are just a couple of options to creating your tables; you can use one or the other, a different method, or a combination.

In [40]:
# Define your database connection parameters
db_connection_string = "postgresql://postgres:1234@localhost:5432/e4501project"
engine = create_engine(db_connection_string)

In [41]:
ZIPCODE_SCHEMA = """
CREATE TABLE IF NOT EXISTS nyc_shape (
  "ZIPCODE" int8 PRIMARY KEY,
  "geometry" geometry(POLYGON, 3857)
);
"""

NYC_311_SCHEMA = """
CREATE TABLE IF NOT EXISTS t311 (
    "Unique Key" int8 PRIMARY KEY,
    "Created Date" timestamp(6),
    "Complaint Type" text COLLATE "pg_catalog"."default",
    "Incident Zip" int8,
    "Latitude" float8,
    "Longitude" float8,
    "Location" text COLLATE "pg_catalog"."default",
    "geometry" geometry(POINT, 3857)
);
"""

NYC_TREE_SCHEMA = """
CREATE TABLE IF NOT EXISTS stc (
    "created_at" timestamp(6),
    "Latitude" float8,
    "longitude" float8,
    "tree_id" int8 PRIMARY KEY,
    "zipcode" int8,
    "status" TEXT,
    "health" text COLLATE "pg_catalog"."default",
    "spc_common" text COLLATE "pg_catalog"."default",
    "geometry" geometry(POINT, 3857)
);
"""

ZILLOW_SCHEMA = """
CREATE TABLE IF NOT EXISTS zillow (
  "index" int8 PRIMARY KEY,
  "RegionID" int8,
  "RegionName" int8,
  "date" DATE,
  "rent" float8
)
;
"""

BEDBUG_SCHEMA = '''
CREATE TABLE IF NOT EXISTS Bedbug (
  "index" int   PRIMARY KEY,
  "Building ID" int8,
  "Postcode" int8,
  "Filing Date" timestamp(6),
  "Eradicated Unit Count" float8,
  "Re-infested  Dwelling Unit Count" float8,
  "Latitude" float8,
  "Longitude" float8,
  "geometry" geometry(POINT,3857)
);
'''

In [42]:
# create that required schema.sql file
with open('schema.sql', "w") as f:
    f.write(ZIPCODE_SCHEMA)
    f.write(NYC_311_SCHEMA)
    f.write(NYC_TREE_SCHEMA)
    f.write(ZILLOW_SCHEMA)
    f.write(BEDBUG_SCHEMA)

In [50]:
!psql --dbname e4501project -f schema.sql

CREATE TABLE
CREATE TABLE
CREATE TABLE
CREATE TABLE
CREATE TABLE


In [49]:
# # Read the SQL schema file
# schema_file_path= "schema.sql"
# with open(schema_file_path, 'r') as file:
#     schema_sql = file.read()
# 
# # Execute the SQL schema
# with engine.connect() as connection:
#     # It's often a good idea to execute commands within a transaction
#     with connection.begin() as transaction:
#         try:
#             connection.execute(schema_sql)
#             # Commit the transaction if all commands execute successfully
#             transaction.commit()
#         except Exception as e:
#             # Rollback the transaction in case of an error
#             transaction.rollback()
#             # Optionally, re-raise the exception or handle it as needed
#             raise e

ObjectNotExecutableError: Not an executable object: '\nCREATE TABLE IF NOT EXISTS nyc_shape (\n  "ZIPCODE" int8 PRIMARY KEY,\n  "geometry" geometry(POLYGON, 3857)\n);\n\nCREATE TABLE IF NOT EXISTS t311 (\n    "Unique Key" int8 PRIMARY KEY,\n    "Created Date" timestamp(6),\n    "Complaint Type" text COLLATE "pg_catalog"."default",\n    "Incident Zip" int8,\n    "Latitude" float8,\n    "Longitude" float8,\n    "Location" text COLLATE "pg_catalog"."default",\n    "geometry" geometry(POINT, 3857)\n);\n\nCREATE TABLE IF NOT EXISTS stc (\n    "created_at" timestamp(6),\n    "Latitude" float8,\n    "longitude" float8,\n    "tree_id" int8 PRIMARY KEY,\n    "zipcode" int8,\n    "status" TEXT,\n    "health" text COLLATE "pg_catalog"."default",\n    "spc_common" text COLLATE "pg_catalog"."default",\n    "geometry" geometry(POINT, 3857)\n);\n\nCREATE TABLE IF NOT EXISTS zillow (\n  "index" int8 PRIMARY KEY,\n  "RegionID" int8,\n  "RegionName" int8,\n  "date" DATE,\n  "rent" float8\n)\n;\n\nCREATE TABLE IF NOT EXISTS Bedbug (\n  "index" int   PRIMARY KEY,\n  "Building ID" int8,\n  "Postcode" int8,\n  "Filing Date" timestamp(6),\n  "Eradicated Unit Count" float8,\n  "Re-infested  Dwelling Unit Count" float8,\n  "Latitude" float8,\n  "Longitude" float8,\n  "geometry" geometry(POINT,3857)\n);\n'

In [51]:
df_zillow_data

index  RegionID  RegionName        date         rent
0          0     62093       11385  2015-01-31          NaN
1          1     62019       11208  2015-01-31          NaN
2          2     62046       11236  2015-01-31          NaN
3          3     61807       10467  2015-01-31          NaN
4          4     62085       11373  2015-01-31          NaN
...      ...       ...         ...         ...          ...
15325  15325     61773       10282  2023-09-30  7347.458333
15326  15326     62010       11109  2023-09-30  4529.358974
15327  15327     61620       10006  2023-09-30  4060.096154
15328  15328     61723       10162  2023-09-30  5011.666667
15329  15329     61618       10004  2023-09-30  4355.328283

[15330 rows x 5 columns]

In [64]:
df_zillow_data.to_sql('zillow', engine, if_exists='append', index=False)
df_bedbug_data.to_postgis('bedbug', engine, if_exists='append', index=False)
geodf_zipcode_data.to_postgis('nyc_shape', engine, if_exists='append', index=False)

In [ ]:
geodf_311_data.to_postgis('t311', engine, if_exists='append', index=False)
geodf_tree_data.to_postgis('stc', engine, if_exists='append', index=False)

## Part 3: Understanding the Data

In [ ]:
# Helper function to write the queries to file
import os


def write_query_to_file(query: str, outfile: str) -> None:
    """
    Writes a given query string to a specified file.

    Args:
    query (str): The query string to write to the file.
    outfile (str): The file path where the query will be written.

    Creates the directory for the outfile if it doesn't exist and writes the query to the file.
    """
    # Create the directory for outfile if it doesn't exist
    os.makedirs(os.path.dirname(outfile), exist_ok=True)

    # Open the file and write the query
    with open(outfile, 'w') as file:
        file.write(query)

# Assert test case
test_query = "SELECT * FROM table;"
test_outfile = "/tmp/test_query.sql"

# Execute the function with the test query and outfile
write_query_to_file(test_query, test_outfile)

# Verify the content of the written file
with open(test_outfile, 'r') as file:
    content = file.read()
assert content == test_query, "Test case failed: The content of the file does not match the query"

# Note: This test case writes to a temporary file. Ensure the file path is appropriate for your environment.





### Query 1

In [ ]:

QUERY_1_FILENAME = "query/area_more_clam.sql"

QUERY_1 = """
SELECT "Incident Zip", COUNT(*) AS number_of_complaints
FROM t311
WHERE "Created Date" BETWEEN '2022-10-01' AND '2023-09-30'
GROUP BY "Incident Zip"
ORDER BY number_of_complaints DESC;
"""
with engine.connect() as conn:
    result = conn.execute(text(QUERY_1))
    for row in result:
        print(row)
write_query_to_file(QUERY_1, QUERY_1_FILENAME)

In conclusion, the area with the zip code 11226 might be more calm to live in, since we found that 11226 has the largest number of 311 complaints in between October 1st, 2022 and September 30th, 2023 (inclusive), which is 48333. 

### Query 2


In [ ]:
QUERY_2_FILENAME = "query/most_greenary.sql"

QUERY_2 = """
SELECT zipcode, COUNT(*) AS total_trees
FROM stc
GROUP BY zipcode
ORDER BY total_trees DESC
LIMIT 10;
"""
with engine.connect() as conn:
    result = conn.execute(text(QUERY_2))
    for row in result:
        print(row)
write_query_to_file(QUERY_2, QUERY_2_FILENAME)

The 10 areas with the zip codes 10312, 10314, 10306, 10309, 11234, 11385, 11357, 11207, 11208, 11434 have the most trees, which are more greenery than the others.

### Query 3


In [ ]:
QUERY_3_FILENAME = "query/most_trees_affordability.sql"

QUERY_3 = """
select zillow."RegionName" , TO_CHAR(AVG(zillow.rent), 'FM999,999,990.00') as avg_price
from zillow inner join (
SELECT zipcode, COUNT(*) AS total_trees
FROM stc
GROUP BY zipcode
ORDER BY total_trees DESC
LIMIT 10
) t on t.zipcode = zillow."RegionName"
where zillow.date >= '2023-08-01'
and zillow.date < '2023-09-01'
group by zillow."RegionName"
order by avg_price desc;

"""
with engine.connect() as conn:
    result = conn.execute(text(QUERY_3))
    for row in result:
        print(row)
write_query_to_file(QUERY_3, QUERY_3_FILENAME)

The average rent costs for area with zip code 11207, 11385, 11208, 11434, 10314, 11357, 10306, 10309, 10312 are 3079.09, 3064.48, 2737.55, 2645.92, 2465.47, 2458.81, 2331.54, 2312.31, 1832.01, 1775.09 respectively.

### Query 4


In [ ]:
QUERY_4_FILENAME = "query/most_trees_affordability.sql"

QUERY_4 = """
(select zillow."RegionName" , avg(zillow.rent) as avg_price , count(tree_id) as tree_numebr
, count(t311."Unique Key") as compaint_number
from zillow
inner join stc on zillow."RegionName" = stc.zipcode
inner join t311 on t311."Incident Zip" = zillow."RegionName"
where zillow.date >= '2023-01-01'
and zillow.date < '2023-02-01'
and t311."Created Date">= '2023-01-01'
and t311."Created Date"< '2023-02-01'
and stc.created_at< '2023-02-01'
group by zillow."RegionName"
HAVING avg(zillow.rent) is not NULL
order by avg_price desc limit 5)
union
(select zillow."RegionName" , avg(zillow.rent) as avg_price , count(tree_id) as tree_numebr
, count(t311."Unique Key") as compaint_number
from zillow
inner join stc on zillow."RegionName" = stc.zipcode
inner join t311 on t311."Incident Zip" = zillow."RegionName"
where zillow.date >= '2023-01-01'
and zillow.date < '2023-02-01'
and t311."Created Date">= '2023-01-01'
and t311."Created Date"< '2023-02-01'
and stc.created_at< '2023-02-01'
group by zillow."RegionName"
HAVING avg(zillow.rent) is not NULL
order by avg_price ASC limit 5)
"""
with engine.connect() as conn:
    result = conn.execute(text(QUERY_3))
    for row in result:
        print(row)
write_query_to_file(QUERY_4, QUERY_4_FILENAME)

For the month of January 2023, the 5 zip codes with the lowest average rent are 10309, 10453, 10462, 11357, 10458 respectively, and 5 zipcodes of the highest average rent are 10013, 10282, 10069, 10011, 10007 respectively, the highest average rents do not correlate directly with the number of trees or complaints.

### Query 5


In [ ]:
QUERY_5_FILENAME = "query/most_greenery_v2.sql"

QUERY_5 = """
SELECT
    nyc_shape."ZIPCODE",
    COUNT(stc.tree_id) as TreeCount
FROM
    nyc_shape
JOIN
    stc
on st_within(stc.geometry,nyc_shape.geometry)
GROUP BY
    nyc_shape."ZIPCODE"
ORDER BY
    TreeCount DESC
LIMIT 10;

"""
with engine.connect() as conn:
    result = conn.execute(text(QUERY_5))
    for row in result:
        print(row)
write_query_to_file(QUERY_5, QUERY_5_FILENAME)

The same as query 2, the 10 areas with the zip codes 10312, 10314, 10306, 10309, 11234, 11385, 11357, 11207, 11208, 11434 have the most trees, which are more greenery than the others.

## query 6

In [ ]:
QUERY_6_FILENAME = "query/immediate_area.sql"

QUERY_6 = """
SELECT
    stc.tree_id AS "ID",
    stc.spc_common AS "species",
    stc.health,
    stc.status,
    ST_AsText(stc.geometry) AS "coordinate location"
FROM
    stc
WHERE
    ST_DWithin(
        stc.geometry,
        ST_Transform(ST_SetSRID(ST_Point(-73.96253174434912, 40.80737875669467), 4326), 3857),
        804.7
    );


"""
with engine.connect() as conn:
    result = conn.execute(text(QUERY_6))
    for row in result:
        print(row)
write_query_to_file(QUERY_6, QUERY_6_FILENAME)

This query summarizes the trees that are within 0.5 mile radius of the point with latitude: 40.80737875669467 and longitude: -73.96253174434912.

## Part 4: Visualizing the Data

### Visulization 1:What can I expect to put up with in NYC?

In [ ]:
import pandas as pd
import psycopg2

def get_data_for_daily_complaints():
    # Database connection parameters
    host = "localhost"
    dbname = "e4501project"
    user = "postgres"
    password = "1234"

    # Connect to the database
    conn = psycopg2.connect(host=host, dbname=dbname, user=user, password=password)

    # Query to find the top 3 complaint types
    top_complaints_query = """
    SELECT
        "Complaint Type"
    FROM
        t311
    WHERE
        "Created Date" >= '2022-10-01' AND "Created Date" <= '2023-09-30'
    GROUP BY
        "Complaint Type"
    ORDER BY
        COUNT("Unique Key") DESC
    LIMIT 3;
    """

    # Execute the top complaints query
    top_complaints_df = pd.read_sql_query(top_complaints_query, conn)
    top_complaints = top_complaints_df['Complaint Type'].tolist()

    # Query to get daily counts for the top 3 complaint types
    daily_counts_query = f"""
    SELECT
        "Complaint Type",
        DATE("Created Date") as Date,
        COUNT("Unique Key") AS Daily_Count
    FROM
        t311
    WHERE
        "Complaint Type" IN %s AND
        "Created Date" >= '2022-10-01' AND
        "Created Date" <= '2023-09-30'
    GROUP BY
        "Complaint Type", DATE("Created Date")
    ORDER BY
        Date;
    """

    # Execute the daily counts query
    daily_counts_df = pd.read_sql_query(daily_counts_query, conn, params=(tuple(top_complaints),))
    daily_counts_df['date'] = pd.to_datetime(daily_counts_df['date'])

    # Close the database connection
    conn.close()

    return daily_counts_df

daily_complaints_df = get_data_for_daily_complaints()

# Assert that the DataFrame is not empty
assert not daily_complaints_df.empty, "DataFrame should not be empty."

# Assert that the DataFrame has the expected columns
expected_columns = {'Complaint Type', 'daily_count', 'date'}
assert set(daily_complaints_df.columns) == set(expected_columns), \
    "DataFrame should contain the columns: 'Complaint Type', 'date', 'Daily_Count'."

# Assert that the column data types are as expected
assert pd.api.types.is_datetime64_any_dtype(daily_complaints_df['date']), \
    "Column 'Date' should be of datetime type."
assert pd.api.types.is_numeric_dtype(daily_complaints_df['daily_count']), \
    "Column 'Daily_Count' should be numeric."

In [ ]:

def create_animation(df: DataFrame) -> FuncAnimation:
    """
    Creates an animated bar chart for daily complaint data.

    Parameters:
        df (DataFrame): A DataFrame containing 'date', 'Complaint Type', and 'daily_count' columns.

    Returns:
        FuncAnimation: An animation object representing the animated bar chart.
    """

    required_columns = {'date', 'Complaint Type', 'daily_count'}
    assert set(df.columns) >= required_columns, "DataFrame lacks required columns for animation."

    fig, ax = plt.subplots(figsize=(12, 6))
    df['date'] = pd.to_datetime(df['date'])
    unique_dates = df['date'].unique()

    def animate(i):
        """
        Update function for the animation, showing cumulative complaints up to each date.
        """
        ax.clear()
        current_date = unique_dates[i]
        current_date_py = pd.to_datetime(current_date).to_pydatetime()

        filtered_df = df[df['date'] <= current_date]
        complaint_counts = filtered_df.groupby('Complaint Type')['daily_count'].sum()

        ax.bar(complaint_counts.index, complaint_counts.values, color=['blue', 'green', 'red'])
        ax.set_xlabel('Complaint Type')
        ax.set_ylabel('Cumulative Number of Complaints')
        ax.set_title(f'Complaints up to {current_date_py.strftime("%Y-%m-%d")}')
        plt.xticks(rotation=0)

    ani = animation.FuncAnimation(fig, animate, frames=len(unique_dates), repeat=False)
    return ani

daily_complaints_df = get_data_for_daily_complaints()
ani = create_animation(daily_complaints_df)

assert isinstance(ani, FuncAnimation), "The function should return a FuncAnimation object."
HTML(ani.to_jshtml())





### Visulization 2:What are the most common complaints in the immediate area?


In [ ]:
import pandas as pd
import psycopg2
from pandas import DataFrame

def get_data_for_complaints_10027() -> DataFrame:
    """
    Retrieves the top 10 complaint types for ZIP code 10027 from a PostgreSQL database.

    The function connects to the database, executes a query to find the top 10 complaint types in the specified ZIP code
    and time frame, and then returns this data as a DataFrame.

    Returns:
        DataFrame: A DataFrame containing the top 10 complaint types and their counts.
    """

    # Database connection parameters
    host = "localhost"
    dbname = "e4501project"
    user = "postgres"
    password = "1234"

    # Connect to the database
    conn = psycopg2.connect(host=host, dbname=dbname, user=user, password=password)

    # Query to find the top 10 complaint types in ZIP code 10027
    complaints_10027_query = """
    SELECT
        "Complaint Type",
        COUNT("Unique Key") AS NumberOfComplaints
    FROM
        t311
    WHERE
        "Incident Zip" = 10027 AND
        "Created Date" BETWEEN '2018-10-01' AND '2023-09-30'
    GROUP BY
        "Complaint Type"
    ORDER BY
        NumberOfComplaints DESC
    LIMIT 10;
    """

    # Execute the query and retrieve the data
    top_complaints_df = pd.read_sql_query(complaints_10027_query, conn)

    # Close the database connection
    conn.close()

    return top_complaints_df

complaints_df = get_data_for_complaints_10027()

# Assert that the DataFrame is not empty
assert not complaints_df.empty, "DataFrame should not be empty."

# Assert that the DataFrame has the expected columns
expected_columns = {'Complaint Type', 'numberofcomplaints'}
assert set(complaints_df.columns) == expected_columns, "DataFrame should have 'Complaint Type' and 'NumberOfComplaints' columns."


In [ ]:
def complaints_10027_figure(df: DataFrame) -> None:
    """
    Generates a horizontal bar chart for the top 10 complaint types in ZIP code 10027.

    Parameters:
        df (DataFrame): A DataFrame containing 'Complaint Type' and 'numberofcomplaints' columns.

    The function creates a bar chart displaying the number of complaints for each complaint type.
    """

    # Assert that the DataFrame has the required columns
    required_columns = {'Complaint Type', 'numberofcomplaints'}
    assert set(df.columns) >= required_columns, "DataFrame lacks required columns for the chart."

    # Creating the bar chart
    plt.figure(figsize=(10, 6))
    plt.barh(df['Complaint Type'], df['numberofcomplaints'], color='skyblue')
    plt.xlabel('Number of Complaints')
    plt.title('Top 10 Complaint Types in Zip Code 10027 (Oct 1, 2018 - Sep 30, 2023)')
    plt.show()

df_10027 = get_data_for_complaints_10027()
complaints_10027_figure(df_10027)

### Visulization 3:  Is there any correlation between rent, trees, and complaints at all?


In [ ]:
import pandas as pd
import psycopg2
from pandas import DataFrame

def get_data_for_correlation() -> DataFrame:
    """
    Retrieves and merges rent, tree, and complaint data from a PostgreSQL database for correlation analysis.

    The function connects to the database, executes queries to retrieve rent data, tree data, and complaint data,
    and then merges these datasets on the 'zipcode' column.

    Returns:
        DataFrame: A DataFrame containing merged rent, tree, and complaint data.
    """

    # Database connection parameters
    host = "localhost"
    dbname = "e4501project"
    user = "postgres"
    password = "1234"

    # Connect to the database
    conn = psycopg2.connect(host=host, dbname=dbname, user=user, password=password)

    # Queries to retrieve data
    rent_data_query = """
        SELECT
            "RegionName" as ZipCode,
            AVG(rent) as AverageRent
        FROM
            zillow
        WHERE
            date >= '2015-01-01' AND date <= '2023-09-30'
        GROUP BY
            "RegionName";
    """

    Tree_Data_query = """
    SELECT
        zipcode,
        COUNT(tree_id) as NumberOfTrees
    FROM
        stc
    GROUP BY
        zipcode;
    """

    Complaint_Data_query = """
    SELECT
        "Incident Zip" as ZipCode,
        COUNT("Unique Key") as NumberOfComplaints
    FROM
        t311
    WHERE
        "Created Date" >= '2015-01-01' AND "Created Date" <= '2023-09-30'
    GROUP BY
        "Incident Zip";
    """

    # Execute the queries and retrieve the data
    rent_data_df = pd.read_sql_query(rent_data_query, conn)
    Tree_Data_df = pd.read_sql_query(Tree_Data_query, conn)
    Complaint_Data_df = pd.read_sql_query(Complaint_Data_query, conn)

    # Merging the data on 'zipcode'
    merged_data = pd.merge(rent_data_df, Tree_Data_df, on='zipcode', how='inner')
    merged_data = pd.merge(merged_data, Complaint_Data_df, on='zipcode', how='inner')

    # Close the database connection
    conn.close()

    return merged_data

correlation_data_df = get_data_for_correlation()

# Assert that the DataFrame is not empty
assert not correlation_data_df.empty, "DataFrame should not be empty."

# Assert that the DataFrame has the expected columns
expected_columns = {'averagerent', 'numberofcomplaints', 'numberoftrees', 'zipcode'}
assert set(correlation_data_df.columns) == expected_columns, "DataFrame should have the expected columns for correlation analysis."

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
from pandas import DataFrame

def genertate_plot(df: DataFrame) -> None:
    """
    Generates scatter plots showing relationships between average rent and number of trees,
    and average rent and number of complaints, by ZIP code.

    Parameters:
        df (DataFrame): A DataFrame containing 'averagerent', 'numberoftrees', and 'numberofcomplaints' columns.
    """

    # Assert that the DataFrame has the required columns
    required_columns = {'averagerent', 'numberoftrees', 'numberofcomplaints'}
    assert set(df.columns) >= required_columns, "DataFrame lacks required columns for the plot."

    # Creating the scatter plots
    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(10, 8), sharex=True)

    # Rent vs Number of Trees
    ax1.scatter(df['averagerent'], df['numberoftrees'], color='green')
    ax1.set_ylabel('Number of Trees')
    ax1.set_title('Rent vs Trees and Complaints by Zip Code')

    # Rent vs Number of Complaints
    ax2.scatter(df['averagerent'], df['numberofcomplaints'], color='red')
    ax2.set_xlabel('Average Rent')
    ax2.set_ylabel('Number of Complaints')

    plt.xticks(rotation=45)
    plt.show()



In [ ]:
merged_data  = get_data_for_correlation()
print(merged_data)

In [ ]:
genertate_plot(merged_data)

### Visulization 4:  If I can afford more in rent, will there be fewer issues & complaints?

In [ ]:
import pandas as pd
import psycopg2
from pandas import DataFrame

def get_merged_data() -> DataFrame:
    """
    Retrieves and merges rent and complaint data from a PostgreSQL database.

    The function connects to the database, executes queries to retrieve rent data and complaint data,
    and then merges these datasets on the 'ZipCode' column.

    Returns:
        DataFrame: A DataFrame containing merged rent and complaint data.
    """

    # Database connection parameters
    host = "localhost"
    dbname = "e4501project"
    user = "postgres"
    password = "1234"

    # SQL queries to fetch rent data and complaint data
    rent_query = """
    SELECT
        "RegionName" AS ZipCode,
        AVG(rent) AS AverageRent
    FROM
        zillow
    WHERE
        date BETWEEN '2023-09-01' AND '2023-09-30'
    GROUP BY
        "RegionName";
    """

    complaints_query = """
    SELECT
        "Incident Zip" AS ZipCode,
        COUNT("Unique Key") AS NumberOfComplaints
    FROM
        t311
    WHERE
        "Created Date" BETWEEN '2022-10-01' AND '2023-09-30'
    GROUP BY
        "Incident Zip";
    """

    # Connect to the database
    conn = psycopg2.connect(host=host, dbname=dbname, user=user, password=password)

    # Execute queries and fetch data
    rent_data = pd.read_sql_query(rent_query, conn)
    complaints_data = pd.read_sql_query(complaints_query, conn)

    # Merge the data on ZipCode
    merged_data = pd.merge(rent_data, complaints_data, on='zipcode', how='inner')

    # Close the database connection
    conn.close()

    return merged_data
merged_df = get_merged_data()

# Assert that the DataFrame is not empty
assert not merged_df.empty, "Merged DataFrame should not be empty."

# Assert that the DataFrame has the expected columns
expected_columns = {'zipcode', 'averagerent', 'numberofcomplaints'}
assert set(merged_df.columns) == expected_columns, "Merged DataFrame should contain 'ZipCode', 'AverageRent', and 'NumberOfComplaints'."


In [ ]:
get_merged_data().columns

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
from pandas import DataFrame

def create_plot(data: DataFrame) -> None:
    """
    Generates a box plot to visualize the relationship between average rent and the number of 311 complaints.

    Parameters:
        data (DataFrame): A DataFrame containing 'averagerent' and 'numberofcomplaints' columns.
    """

    # Assert that the DataFrame has the required columns
    required_columns = {'averagerent', 'numberofcomplaints'}
    assert set(data.columns) >= required_columns, "DataFrame lacks required columns for the plot."

    # Creating rent bins for categorization
    rent_bins = range(0, int(data['averagerent'].max()) + 1000, 1000)
    rent_labels = [f"${i}-${i + 999}" for i in rent_bins[:-1]]
    data['RentBin'] = pd.cut(data['averagerent'], bins=rent_bins, labels=rent_labels, right=False)

    # Preparing data for the boxplot
    boxplot_data = [group['numberofcomplaints'].values for _, group in data.groupby('RentBin')]

    # Creating the boxplot
    plt.figure(figsize=(12, 6))
    plt.boxplot(boxplot_data, labels=rent_labels)
    plt.xlabel('Average Rent (in bins)')
    plt.ylabel('Number of 311 Complaints')
    plt.title('Number of 311 Complaints vs. Average Rent by Zip Code (Oct 1, 2022 - Sep 30, 2023)')
    plt.xticks(rotation=45)
    plt.show()



In [ ]:
# Call the function to create the plot
merged_data = get_merged_data()
create_plot(merged_data)

### Visulization 5:  Where were the recent 311 incidents reported from in the immediate area?


In [ ]:
import pandas as pd
import psycopg2
from pandas import DataFrame

def get_311_incidents_data() -> DataFrame:
    """
    Retrieves 311 incident data (latitude and longitude) from a PostgreSQL database.

    The function connects to the database, executes a query to retrieve incident data
    within a specific radius of a given point, and then returns this data as a DataFrame.

    Returns:
        DataFrame: A DataFrame containing the latitude and longitude of 311 incidents.
    """

    # Database connection parameters
    host = "localhost"
    dbname = "e4501project"
    user = "postgres"
    password = "1234"

    # SQL query to fetch 311 incident data
    incidents_query = """
    SELECT
        "Latitude",
        "Longitude"
    FROM
        t311
    WHERE
        "Created Date" BETWEEN '2023-01-01' AND '2023-09-30'
        AND ST_DWithin(
            ST_MakePoint("Longitude", "Latitude")::geography,
            ST_MakePoint(-73.96253174434912, 40.80737875669467)::geography,
            1000
        );
    """

    # Connect to the database
    conn = psycopg2.connect(host=host, dbname=dbname, user=user, password=password)

    # Execute query and fetch data
    incidents_data = pd.read_sql_query(incidents_query, conn)

    # Close the database connection
    conn.close()

    return incidents_data

incidents_df = get_311_incidents_data()

# Assert that the DataFrame is not empty
assert not incidents_df.empty, "DataFrame should not be empty."

# Assert that the DataFrame has the expected columns
expected_columns = {'Latitude', 'Longitude'}
assert set(incidents_df.columns) == expected_columns, "DataFrame should contain 'Latitude' and 'Longitude' columns."


In [ ]:
import pandas as pd
import geopandas as gpd
import matplotlib.pyplot as plt
import contextily as ctx
from shapely.geometry import Point
from pandas import DataFrame
from geopandas import GeoDataFrame

def plot_geospatial_data(df: DataFrame) -> None:
    """
    Plots geospatial data from a DataFrame containing Latitude and Longitude columns.

    Parameters:
    df (DataFrame): DataFrame containing the geospatial data with 'Latitude' and 'Longitude'.
    """

    # Assert that the DataFrame has the required columns
    required_columns = {'Latitude', 'Longitude'}
    assert set(df.columns) >= required_columns, "DataFrame lacks required columns for geospatial plotting."

    # Convert DataFrame to GeoDataFrame with WGS 84 (EPSG:4326) CRS
    gdf = gpd.GeoDataFrame(
        df,
        geometry=[Point(xy) for xy in zip(df.Longitude, df.Latitude)],
        crs="EPSG:4326"  # WGS 84 coordinate reference system
    )

    # Convert GeoDataFrame to Web Mercator (EPSG:3857) for mapping
    gdf = gdf.to_crs(epsg=3857)

    # Plotting
    fig, ax = plt.subplots(figsize=(10, 10))
    gdf.plot(ax=ax, color='red', markersize=5)

    # Adding a basemap using contextily
    ctx.add_basemap(ax, source=ctx.providers.OpenStreetMap.Mapnik, zoom='auto')

    # Setting axis labels and adjusting axis limits
    ax.set_xlabel('Longitude')
    ax.set_ylabel('Latitude')
    ax.set_xlim(gdf.total_bounds[[0, 2]])
    ax.set_ylim(gdf.total_bounds[[1, 3]])

    # Adding gridlines and title to the plot
    ax.grid(True)
    ax.set_title('Reported 311 Incidents (Jan 1, 2023 - Sep 30, 2023) within 1km Radius')

    plt.show()

In [ ]:
merge_data = get_311_incidents_data()
merge_data
plot_geospatial_data(merge_data)

### Visulization 6:  Are areas trying to improve the amount of trees in the neighborhood?


In [ ]:
import pandas as pd
import psycopg2
from pandas import DataFrame
from typing import Tuple

def get_trees_and_complaints_data() -> Tuple[DataFrame, DataFrame]:
    """
    Retrieves trees data and 311 complaints data related to 'New Tree Request' from a PostgreSQL database.

    Returns:
        Tuple[DataFrame, DataFrame]: Two DataFrames containing trees data and 311 complaints data.
    """

    # Database connection parameters
    host = "localhost"
    dbname = "e4501project"
    user = "postgres"
    password = "1234"

    # SQL query to fetch trees data
    trees_query = """
    SELECT
        "Latitude",
        "longitude"
    FROM
        stc;
    """

    # SQL query to fetch 311 complaints data
    complaints_query = """
    SELECT
        "Latitude",
        "Longitude"
    FROM
        t311
    WHERE
        "Complaint Type" = 'New Tree Request'
        AND "Created Date" BETWEEN '2018-10-01' AND '2023-09-30'
        AND "Latitude" IS NOT NULL
        AND "Longitude" IS NOT NULL;
    """

    # Connect to the database
    conn = psycopg2.connect(host=host, dbname=dbname, user=user, password=password)

    # Execute queries and fetch data
    trees_data = pd.read_sql_query(trees_query, conn)
    complaints_data = pd.read_sql_query(complaints_query, conn)

    # Close the database connection
    conn.close()

    return trees_data, complaints_data

trees_df, complaints_df = get_trees_and_complaints_data()

# Assert that both DataFrames are not empty
assert not trees_df.empty, "Trees DataFrame should not be empty."
assert not complaints_df.empty, "Complaints DataFrame should not be empty."

# Assert that both DataFrames have the expected columns
expected_columns_tree = {'Latitude', 'longitude'}
assert set(trees_df.columns) == expected_columns_tree, "Trees DataFrame should contain 'Latitude' and 'Longitude'."

expected_columns = {'Latitude', 'Longitude'}
assert set(complaints_df.columns) == expected_columns, "Complaints DataFrame should contain 'Latitude' and 'Longitude'."


In [ ]:
import pandas as pd
import geopandas as gpd
import matplotlib.pyplot as plt
import contextily as ctx
from shapely.geometry import Point
from pandas import DataFrame
from geopandas import GeoDataFrame

def plot_geospatial_data(trees_df: DataFrame, complaints_df: DataFrame) -> None:
    """
    Plots geospatial data for trees and 311 new tree requests with a basemap.

    Parameters:
    trees_df (DataFrame): DataFrame containing tree data with 'Latitude' and 'Longitude'.
    complaints_df (DataFrame): DataFrame containing 311 complaints data with 'Latitude' and 'Longitude'.
    """

    # Assert that the DataFrames have the required columns
    assert all(col in trees_df.columns for col in ['Latitude', 'longitude']), \
        "Trees DataFrame lacks required columns 'Latitude' and 'Longitude'."
    assert all(col in complaints_df.columns for col in ['Latitude', 'Longitude']), \
        "Complaints DataFrame lacks required columns 'Latitude' and 'Longitude'."

    # Convert DataFrames to GeoDataFrames with WGS 84 (EPSG:4326) CRS
    gdf_trees = gpd.GeoDataFrame(
        trees_df,
        geometry=[Point(xy) for xy in zip(trees_df.longitude, trees_df.Latitude)],
        crs="EPSG:4326"
    )
    gdf_complaints = gpd.GeoDataFrame(
        complaints_df,
        geometry=[Point(xy) for xy in zip(complaints_df.Longitude, complaints_df.Latitude)],
        crs="EPSG:4326"
    )

    # Convert GeoDataFrames to Web Mercator (EPSG:3857) for mapping
    gdf_trees = gdf_trees.to_crs(epsg=3857)
    gdf_complaints = gdf_complaints.to_crs(epsg=3857)

    # Plotting
    fig, ax = plt.subplots(figsize=(12, 12))
    gdf_trees.plot(ax=ax, color='green', markersize=5, label='Trees')
    gdf_complaints.plot(ax=ax, color='red', markersize=5, label='New Tree Requests')

    # Adding a basemap using contextily
    ctx.add_basemap(ax, source=ctx.providers.OpenStreetMap.Mapnik)

    # Setting axis labels and adding a legend
    ax.set_xlabel('Longitude')
    ax.set_ylabel('Latitude')
    ax.set_title('NYC Trees and New Tree Requests (Oct 1, 2018 - Sep 30, 2023)')
    ax.legend()

    plt.show()

In [ ]:
tree,complaints_data = get_trees_and_complaints_data()

In [ ]:
plot_geospatial_data(tree,complaints_data)